docs/src/dev/future/proposal-multi-threaded-transaction-5.asciidoc

```diff
@@ -0,0 +1,146 @@
+At the stage of resolving bindings for Graph and GraphTraversalSource need to create an instance that will be specific only for the current session/tx and use it until the end of the operation ending with commit/rollback. Now bindings are stored in different places, for example GremlinScriptEngineManager.globalScope and GraphManager.getTraversalSource/getGraph, also need to add a used Graph and GraphTraversalSource to Session.
+
+
+==== Overview of changes to code base
+===== TinkerTransactionGraph2 (suggest a better name)
+Storage for committed data. Should not be used directly for graph operations.
+[code]
+----
+class TinkerTransactionGraph2 implements Graph {
+    private final Map<Object, TinkerElementLockContainer<TinkerVertex>> vertices = new ConcurrentHashMap<>();
+
+    // I haven't decided which option is better yet
+    public Transaction tx() { throw new Exception("use TinkerGraphTx instead"); }
+    public Transaction tx() { return new TinkerGraphTx(this); }
+}
+----
+
+===== TinkerElementLockContainer
+Storage for Element and lock for concurrent modification.
+[code]
+----
+class TinkerElementLockContainer<T extends TinkerElement> {
+    T element;
+	ReentrantLock lock = new ReentrantLock();
+}
+----
+
+===== TinkerGraphTx (suggest a better name)
+Graph for use in single transaction
+[code]
+----
+class TinkerGraphTx implements Graph, Transaction {
+    private final Map<Object, TinkerElementTxContainer<TinkerVertex>> changedVertices = new ConcurrentHashMap<>();
+	private final TinkerTransactionGraph2 commitedGraph;
+
+    public Transaction tx() { return this; }
+
+	public Vertex addVertex(final Object... keyValues) {
+		// create and add vertex to changedVertices
+	}
+
+	public Vertex vertex(final Object vertexId) {
+		// search for vertex in changedVertices, if not found than in commitedGraph
+	}
+
+	protected void doCommit() throws TransactionException {
+		// propagate changes to commitedGraph
+	}
+}
+----
+
+===== TinkerElementTxContainer (suggest a better name)
+Storage for element used in transaction.
+Can be unmodified (cached for consistent read) or created/modified in current tx.
+Unchanged element is simply a reference to the element that was there at the time of first reading. When a change
+is made, a clone is created and further work is done with it.
+[code]
+----
+final class TinkerElementTxContainer<T extends TinkerElement> {
+	T element;
+	boolean isDeleted;
+	boolean isModified;
+}
+----
+
+
+==== CRUD operations without transaction
+1. Create TinkerGraphTx and GraphTraversalSource from it.
+2. Execute traversal
+3. Commit or rollback on error.
+
+
+==== CRUD operations with transaction
+===== Create
+Add new TinkerElementTxContainer to `vertices` and `edges` in TinkerGraphTx.
+
+==== Read
+Read values from `vertices` TinkerElementTxContainer in TinkerGraphTx.
+If marked as deleted than return null;
+If not found try TinkerTransactionGraph2.vertices.
```

Comment:
Would we want to add the following to ensure consistent reads?

If found in `TinkerTransactionGraph2.vertices`, copy to `vertices` TinkerElementTxContainer in TinkerGraphTx?